File: src/entities/dataplane.ts
import { JsonLdId, JsonLdObject, JsonLdValue } from "./jsonld";

export class Dataplane extends JsonLdId {

  get url(): string {
    return this.mandatoryValue('edc', 'url');
  }

  get allowedSourceTypes(): String[] {
    return this.arrayOf(() => new JsonLdValue<string>(), 'edc', 'allowedSourceTypes')
      .map(it => it.value);
  }

  get allowedDestTypes(): String[] {
    return this.arrayOf(() => new JsonLdValue<string>(), 'edc', 'allowedDestTypes')
      .map(it => it.value);
  }

  get properties(): JsonLdObject {
    return this.nested('edc', 'properties');
  }

}

export type DataplaneInput = Omit<Dataplane, "lastActive" | "turnCount">;

File: rust/src/entities/dataplane.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Dataplane {
    pub id: String,
    pub url: String,
    pub last_active: u64,
    pub turn_count: u64,
    pub allowed_source_types: Vec<String>,
    pub allowed_dest_types: Vec<String>,
    pub properties: HashMap<String, String>,
}

#[derive(Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DataplaneInput {
    pub id: String,
    pub url: String,
    pub allowed_source_types: Vec<String>,
    pub allowed_dest_types: Vec<String>,
    pub properties: HashMap<String, String>,
}
